Mappings when one project depends upon another

When one project depends upon another, the OwlModels/ont-policy.rdf mapping file does not contain the mappings necessary to handle imports correctly. It currently works (at least in some cases) because Jena caches models and provides them. But other things that depend on correct dependencies (e.g., import graph) will not be correct.

@kittaakos , when I try to build on this branch I get this error:
`[ERROR] Failures:
[ERROR]   SadlRenameTest.testRenameConflict:254->testRename:359 Expceted a failure with message: 'Refactoring cannot be performed'.
[ERROR]   SadlRenameTest.testRenameDecl_MultiFile:229->testRename:382->assertEquals:259->AbstractLanguageServerTest.assertEquals:1398 expected:<changes :
[    Other.sadl : Bar [[2, 11] .. [2, 14]]
    TestMe.sadl : Bar [[1, 0] .. [1, 3]]
`
The error appears to be in the language server:
`				languageServer.rename(new RenameParams => [
					textDocument = new TextDocumentIdentifier(uri)
					position = new Position(line, column)
					it.newName = configuration.newName
				]).get
				fail('''Expceted a failure with message: '«configuration.error»'.''')
`.
Was this expected?

--- a/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl.applications/src/com/ge/research/sadl/applications/SADLCliAppRuntimeModule.java
+++ b/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl.applications/src/com/ge/research/sadl/applications/SADLCliAppRuntimeModule.java
@@ -29,6 +29,7 @@ import java.io.File;
 import java.io.IOException;
 import java.net.URI;
 import java.nio.file.Path;
+import java.util.List;
 import java.util.Map;
 import org.eclipse.emf.common.notify.Notifier;
 import org.eclipse.emf.ecore.resource.ResourceSet;
@@ -131,6 +132,16 @@ public class SADLCliAppRuntimeModule extends SADLRuntimeModule {
                 return eclipseHelper.toPath(uri);
             }
         }
+
+        @Override
+        public List<URI> getReferencedProjectURIs(URI uri) {
+            if (usesFileScheme(uri)) {
+                return ideHelper.getReferencedProjectURIs(uri);
+            } else {
+                return eclipseHelper.getReferencedProjectURIs(uri);
+            }
+        }
+
     }
 
     public Class<? extends SadlGraphVisualizerHandler> bindSadlGraphVisualizerHandler() {

--- a/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl.jena/src/com/ge/research/sadl/jena/JenaBasedSadlModelProcessor.java
+++ b/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl.jena/src/com/ge/research/sadl/jena/JenaBasedSadlModelProcessor.java
@@ -526,10 +526,10 @@ public class JenaBasedSadlModelProcessor extends SadlModelProcessor implements I
 	@Override
 	public void onGenerate(Resource resource, IFileSystemAccess2 fsa, ProcessorContext context) {
 		// update project dependencies
-		List<java.net.URI> projectDependencies = projectHelper.getReferencedProjectURIs(resource);
 		try {
+			List<java.net.URI> projectDependencies = projectHelper.getReferencedProjectURIs(resource);
 			getConfigMgr().addProjectDependencies(projectDependencies);
-		} catch (ConfigurationException e2) {
+		} catch (Exception e2) {
 			// TODO Auto-generated catch block
 			e2.printStackTrace();
 		}

--- a/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl.jena/src/com/ge/research/sadl/jena/JenaBasedSadlModelProcessor.java
+++ b/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl.jena/src/com/ge/research/sadl/jena/JenaBasedSadlModelProcessor.java
@@ -525,7 +525,12 @@ public class JenaBasedSadlModelProcessor extends SadlModelProcessor implements I
 	 */
 	@Override
 	public void onGenerate(Resource resource, IFileSystemAccess2 fsa, ProcessorContext context) {
-				
+		// test ability to get project dependencies
+		List<java.net.URI> projectDependencies = projectHelper.getReferencedProjectURIs(resource);
+		for (java.net.URI pduri : projectDependencies) {
+			System.err.println("Project '" + resource.getURI().toString() + "' depends on '" + pduri.toString());
+		}
+		
 		generationInProgress = true;
 		setCurrentResource(resource);
 		setProcessorContext(context);

--- a/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl.jena/src/com/ge/research/sadl/jena/JenaBasedSadlModelProcessor.java
+++ b/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl.jena/src/com/ge/research/sadl/jena/JenaBasedSadlModelProcessor.java
@@ -525,10 +525,13 @@ public class JenaBasedSadlModelProcessor extends SadlModelProcessor implements I
 	 */
 	@Override
 	public void onGenerate(Resource resource, IFileSystemAccess2 fsa, ProcessorContext context) {
-		// test ability to get project dependencies
+		// update project dependencies
 		List<java.net.URI> projectDependencies = projectHelper.getReferencedProjectURIs(resource);
-		for (java.net.URI pduri : projectDependencies) {
-			System.err.println("Project '" + resource.getURI().toString() + "' depends on '" + pduri.toString());
+		try {
+			getConfigMgr().addProjectDependencies(projectDependencies);
+		} catch (ConfigurationException e2) {
+			// TODO Auto-generated catch block
+			e2.printStackTrace();
 		}
 		
 		generationInProgress = true;

--- a/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl.reasoner/com.ge.research.sadl.reasoner-api/src/main/java/com/ge/research/sadl/reasoner/IConfigurationManager.java
+++ b/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl.reasoner/com.ge.research.sadl.reasoner-api/src/main/java/com/ge/research/sadl/reasoner/IConfigurationManager.java
@@ -301,9 +301,16 @@ public interface IConfigurationManager {
 	public abstract String fileUrlToFileName(String url) throws MalformedURLException;
 	
 	/**
-	 * Metod to convert a File-compatible tring to a file URL string
+	 * Method to convert a File-compatible tring to a file URL string
 	 * @param fn
 	 * @return
 	 */
 	public abstract String fileNameToFileUrl(String fn);
+
+	/**
+	 * Method to get a list of projects (kbases) upon which the project (kbase) for this ConfigurationManager depends
+	 * @return
+	 * @throws ConfigurationException
+	 */
+	public abstract List<String> getProjectDependencies() throws ConfigurationException;
 }

--- a/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl.reasoner/com.ge.research.sadl.reasoner-impl/src/main/java/com/ge/research/sadl/reasoner/ConfigurationManager.java
+++ b/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl.reasoner/com.ge.research.sadl.reasoner-impl/src/main/java/com/ge/research/sadl/reasoner/ConfigurationManager.java
@@ -1567,15 +1567,33 @@ public class ConfigurationManager implements IConfigurationManager {
 	public static final String[] PROJECT_DEPENDENCIES_CATEGORY = {"ProjectDependencies"};
 	public static final String pdependsOn = "pDependsOn";
 
-	private String getOtherProjectAltUrlFromPubliceUri(String publicUri) throws ConfigurationException {
-		String[] categoryHierarchy = PROJECT_DEPENDENCIES_CATEGORY;
-		List<ConfigurationItem> configItems = getConfiguration(categoryHierarchy, false);
-		if (configItems != null && configItems.size() > 0) {
-			ConfigurationItem configItem = configItems.get(0);
-			String dp = configItem.getNamedValue(pdependsOn).toString();
-			
-			IConfigurationManager dpCM = ConfigurationManagerFactory.getConfigurationManager(dp + "/OwlModels", null);
-			return dpCM.getAltUrlFromPublicUri(publicUri);
+	@Override
+	public List<String> getProjectDependencies() throws ConfigurationException {
+		List<ConfigurationItem> config = getConfiguration(PROJECT_DEPENDENCIES_CATEGORY, false);
+		List<Object> previousQueries = null;
+		if (config != null && config.size() > 0) {
+			previousQueries  = config.get(0).getAllValuesOfName(pdependsOn);
+		}
+		if (previousQueries != null && previousQueries.size() >= 1) {
+			List<String> dependsOnProjects = new ArrayList<String>();
+			for (Object dependsOn : previousQueries) {
+				dependsOnProjects.add(dependsOn.toString());
+			}
+			return dependsOnProjects;
+		}
+		return null;
+	}
+
+	protected String getOtherProjectAltUrlFromPubliceUri(String publicUri) throws ConfigurationException {
+		List<String> dps = getProjectDependencies();
+		if (dps != null) {
+			for (String dp : dps) {
+				IConfigurationManager dpCM = ConfigurationManagerFactory.getConfigurationManager(dp + "/OwlModels", null);
+				String dpAltUrl = dpCM.getAltUrlFromPublicUri(publicUri);
+				if (dpAltUrl != null) {
+					return dpAltUrl;
+				}
+			}
 		}
 		return null;
 	}

--- a/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl.reasoner/com.ge.research.sadl.reasoner-impl/src/main/java/com/ge/research/sadl/reasoner/ConfigurationManager.java
+++ b/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl.reasoner/com.ge.research.sadl.reasoner-impl/src/main/java/com/ge/research/sadl/reasoner/ConfigurationManager.java
@@ -1548,12 +1548,38 @@ public class ConfigurationManager implements IConfigurationManager {
 		if (jdm != null) {
 			String alt = getJenaDocumentMgr().doAltURLMapping(publicUri);
 			if (alt != null) {
+				if (alt.equals(publicUri)) {
+					String otherProjectAlt = getOtherProjectAltUrlFromPubliceUri(publicUri);
+					if (otherProjectAlt != null) {
+						return otherProjectAlt;
+					}
+				}
 				return alt;
 			}
+		}
+		String otherProjectAlt = getOtherProjectAltUrlFromPubliceUri(publicUri);
+		if (otherProjectAlt != null) {
+			return otherProjectAlt;
 		}
 		throw new ConfigurationException("PublicURI '" + publicUri + "' not found in mappings.");
 	}
 	
+	public static final String[] PROJECT_DEPENDENCIES_CATEGORY = {"ProjectDependencies"};
+	public static final String pdependsOn = "pDependsOn";
+
+	private String getOtherProjectAltUrlFromPubliceUri(String publicUri) throws ConfigurationException {
+		String[] categoryHierarchy = PROJECT_DEPENDENCIES_CATEGORY;
+		List<ConfigurationItem> configItems = getConfiguration(categoryHierarchy, false);
+		if (configItems != null && configItems.size() > 0) {
+			ConfigurationItem configItem = configItems.get(0);
+			String dp = configItem.getNamedValue(pdependsOn).toString();
+			
+			IConfigurationManager dpCM = ConfigurationManagerFactory.getConfigurationManager(dp + "/OwlModels", null);
+			return dpCM.getAltUrlFromPublicUri(publicUri);
+		}
+		return null;
+	}
+
 	/**
 	 * Method to determine if a URI (could be public URI or alt URL) is mapped in this ConfigurationManager (project)
 	 * @param uri

--- a/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl.reasoner/com.ge.research.sadl.reasoner-impl/src/main/java/com/ge/research/sadl/reasoner/ConfigurationManagerForEditing.java
+++ b/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl.reasoner/com.ge.research.sadl.reasoner-impl/src/main/java/com/ge/research/sadl/reasoner/ConfigurationManagerForEditing.java
@@ -1154,6 +1154,7 @@ public class ConfigurationManagerForEditing extends ConfigurationManager
 	 */
 	public void updateConfiguration(ConfigurationItem newItem)
 			throws ConfigurationException {
+		boolean bChanged = false;
 		if (getConfigModel() != null) {
 			Resource type = getConfigModel().createResource(CATEGORY_KW);
 			Property subcatprop = getConfigModel().createProperty(
@@ -1166,6 +1167,7 @@ public class ConfigurationManagerForEditing extends ConfigurationManager
 						CONFIG_NAMESPACE + categoryHierarchy[i], type);
 				if (i > 0) {
 					getConfigModel().add(lastCategory, subcatprop, newCategory);
+					bChanged = true;
 				}
 				lastCategory = newCategory;
 			}
@@ -1191,6 +1193,7 @@ public class ConfigurationManagerForEditing extends ConfigurationManager
 															+ nvp.getName()));
 							updateStmt.changeObject(getConfigModel()
 									.createTypedLiteral(nvp.getValue()));
+							bChanged = true;
 						}
 					} else {
 						getConfigModel().add(
@@ -1198,11 +1201,14 @@ public class ConfigurationManagerForEditing extends ConfigurationManager
 								prop,
 								getConfigModel().createTypedLiteral(
 										nvp.getValue()));
+						bChanged = true;
 					}
 				}
 			}
-			clearReasoner();
-			setConfigChanged(true);
+			if (bChanged) {
+				clearReasoner();
+				setConfigChanged(true);
+			}
 		}
 	}
 

--- a/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl.reasoner/com.ge.research.sadl.reasoner-impl/src/main/java/com/ge/research/sadl/reasoner/SadlJenaModelGetter.java
+++ b/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl.reasoner/com.ge.research.sadl.reasoner-impl/src/main/java/com/ge/research/sadl/reasoner/SadlJenaModelGetter.java
@@ -29,6 +29,7 @@ import java.net.MalformedURLException;
 import java.net.URL;
 import java.net.URLConnection;
 import java.util.Iterator;
+import java.util.List;
 
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
@@ -192,6 +193,18 @@ public class SadlJenaModelGetter implements ModelGetter, ISadlJenaModelGetter {
     	}
     	if (m == null && loadIfAbsent != null) {
             String altUrl = configurationManager.getJenaDocumentMgr().doAltURLMapping(uri);
+            if (altUrl != null && altUrl.equals(uri)) {
+            	// try for a dependent project
+            	try {
+					String dpAltUrl = configurationManager.getAltUrlFromPublicUri(uri);
+					if (dpAltUrl != null) {
+						altUrl = dpAltUrl;
+					}
+				} catch (ConfigurationException e) {
+					// TODO Auto-generated catch block
+					e.printStackTrace();
+				}
+            }
             if (altUrl != null && altUrl.endsWith(".TDB/")) {
             	try {
             		SadlUtils su = new SadlUtils();

--- a/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl.ui.tests/src/com/ge/research/sadl/ui/tests/AbstractSadlPlatformTest.xtend
+++ b/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl.ui.tests/src/com/ge/research/sadl/ui/tests/AbstractSadlPlatformTest.xtend
@@ -137,12 +137,19 @@ abstract class AbstractSadlPlatformTest extends Assert {
 	 */
 	protected def void createProject() {
 		createProject(projectName);
+	}
+
+	/**
+	 * Creates the test project with the given name.
+	 */
+	protected def void createProject(String projectName) {
+		val project = IResourcesSetupUtil.createProject(projectName);
 		addNature(project, XtextProjectHelper.NATURE_ID);
 		addBuilder(project, XtextProjectHelper.BUILDER_ID);
 		waitForBuild;
 		configurePreferences();
 		// This is used to trigger the implicit model creation before the tests.
-		val file = createFile('Dummy.sadl', 'uri "http://sadl.org/Dummy.sadl."');
+		val file = createFile(project.name, 'Dummy.sadl', 'uri "http://sadl.org/Dummy.sadl."');
 		fullBuild();
 		file.delete(true, monitor);
 		fullBuild();
@@ -268,9 +275,13 @@ abstract class AbstractSadlPlatformTest extends Assert {
 	}
 
 	/**
-	 * Creates a file with the given file name and content. 
+	 * Creates a file with the given file name and content.
 	 */
 	protected def createFile(String fileName, String content) {
+		createFile(projectName, fileName, content)
+	}
+
+	protected def createFile(String projectName, String fileName, String content) {
 		val file = IResourcesSetupUtil.createFile('''«projectName»«SEPARATOR»«fileName»''', content);
 		assertNotNull('''Cannot create '«fileName»' file in the workspace.''', file);
 		assertTrue('''Cannot access '«fileName»' file project in the workspace.''', file.accessible);

--- a/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl.ui.tests/src/com/ge/research/sadl/ui/tests/GH_523_ProjectDependenciesUITest.xtend
+++ b/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl.ui.tests/src/com/ge/research/sadl/ui/tests/GH_523_ProjectDependenciesUITest.xtend
@@ -1,0 +1,94 @@
+package com.ge.research.sadl.ui.tests
+
+import com.ge.research.sadl.utils.SadlProjectHelper
+import com.google.inject.Inject
+import java.net.URI
+import java.util.UUID
+import org.eclipse.core.resources.IResource
+import org.junit.Test
+
+class GH_523_ProjectDependenciesUITest extends AbstractSadlPlatformTest {
+
+	@Inject
+	SadlProjectHelper testMe
+
+	@Test
+	def void testCanReadEmpty() {
+		val file = createFile('Foo.sadl', 'uri "http://sadl.org/Foo.sadl".')
+		assertEquals(0, testMe.getReferencedProjectURIs(file.toUri).length)
+		assertEquals(0, testMe.getReferencedProjectURIs(file.project.toUri).length)
+		assertEquals(0, testMe.getReferencedProjectURIs(file.resource).length)
+	}
+
+	@Test
+	def void testOneProject() {
+		val otherProjectName = '''sadl--gh-523--test-other-project-«UUID.randomUUID»'''
+		createProject(otherProjectName)
+		val otherProject = projects.findFirst[name == otherProjectName]
+		assertTrue(otherProject.accessible)
+
+		val copy = project.description
+		copy.referencedProjects = #[otherProject]
+		project.setDescription(copy, null)
+
+		val file = createFile('Foo.sadl', 'uri "http://sadl.org/Foo.sadl".')
+		assertEquals(1, testMe.getReferencedProjectURIs(file.toUri).length)
+		assertEquals(1, testMe.getReferencedProjectURIs(file.project.toUri).length)
+		assertEquals(1, testMe.getReferencedProjectURIs(file.resource).length)
+	}
+
+	@Test
+	def void testMultiProject() {
+		val projectOneName = '''sadl--gh-523--test-project-one-«UUID.randomUUID»'''
+		createProject(projectOneName)
+		val projectOne = projects.findFirst[name == projectOneName]
+		assertTrue(projectOne.accessible)
+
+		val projectTwoName = '''sadl--gh-523--test-project-two-«UUID.randomUUID»'''
+		createProject(projectTwoName)
+		val projectTwo = projects.findFirst[name == projectTwoName]
+		assertTrue(projectTwo.accessible)
+
+		val copy = project.description
+		copy.referencedProjects = #[projectOne, projectTwo]
+		project.setDescription(copy, null)
+
+		val file = createFile('Foo.sadl', 'uri "http://sadl.org/Foo.sadl".')
+		assertEquals(2, testMe.getReferencedProjectURIs(file.toUri).length)
+		assertEquals(2, testMe.getReferencedProjectURIs(file.project.toUri).length)
+		assertEquals(2, testMe.getReferencedProjectURIs(file.resource).length)
+		assertEquals(#[projectOne.locationURI, projectTwo.locationURI].sort,
+			testMe.getReferencedProjectURIs(file.resource).sort)
+	}
+
+	@Test
+	def void testMultiProjectWithMissing() {
+		val projectOneName = '''sadl--gh-523--test-project-one-«UUID.randomUUID»'''
+		createProject(projectOneName)
+		val projectOne = projects.findFirst[name == projectOneName]
+		assertTrue(projectOne.accessible)
+
+		val projectTwoName = '''sadl--gh-523--test-project-two-«UUID.randomUUID»'''
+		createProject(projectTwoName)
+		val projectTwo = projects.findFirst[name == projectTwoName]
+		assertTrue(projectTwo.accessible)
+
+		val copy = project.description
+		copy.referencedProjects = #[projectOne, projectTwo]
+		project.setDescription(copy, null)
+
+		projectOne.delete(true, null)
+		assertFalse(projectOne.accessible)
+
+		val file = createFile('Foo.sadl', 'uri "http://sadl.org/Foo.sadl".')
+		assertEquals(1, testMe.getReferencedProjectURIs(file.toUri).length)
+		assertEquals(1, testMe.getReferencedProjectURIs(file.project.toUri).length)
+		assertEquals(1, testMe.getReferencedProjectURIs(file.resource).length)
+		assertEquals(#[projectTwo.locationURI], testMe.getReferencedProjectURIs(file.resource))
+	}
+
+	protected def URI toUri(IResource resource) {
+		return testMe.toUri(resource.location.toFile.toPath)
+	}
+
+}

--- a/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl.ui/src/com/ge/research/sadl/ui/visualize/GraphGenerator.java
+++ b/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl.ui/src/com/ge/research/sadl/ui/visualize/GraphGenerator.java
@@ -39,9 +39,12 @@ import com.ge.research.sadl.model.visualizer.IGraphVisualizer;
 import com.ge.research.sadl.preferences.SadlPreferences;
 import com.ge.research.sadl.processing.SadlConstants;
 import com.ge.research.sadl.reasoner.ConfigurationException;
+import com.ge.research.sadl.reasoner.ConfigurationItem;
 import com.ge.research.sadl.reasoner.IConfigurationManagerForEditing.Scope;
 import com.ge.research.sadl.reasoner.InvalidNameException;
 import com.ge.research.sadl.reasoner.ResultSet;
+import com.ge.research.sadl.reasoner.ConfigurationItem.NameValuePair;
+import com.ge.research.sadl.reasoner.ConfigurationManager;
 import com.ge.research.sadl.reasoner.utils.SadlUtils;
 import com.ge.research.sadl.sADL.SadlResource;
 import com.ge.research.sadl.ui.handlers.SadlActionHandler;
@@ -172,7 +175,19 @@ public class GraphGenerator {
 		setVisualizer(visualizer);
 		setProject(project);
 		if (publicUri != null) {
-			setTheJenaModel(configMgr.getOntModel(publicUri, Scope.LOCALONLY));
+			boolean processImports = false;
+			String[] categoryHierarchy = ConfigurationManager.PROJECT_DEPENDENCIES_CATEGORY;
+			List<ConfigurationItem> configItems = getConfigMgr().getConfiguration(categoryHierarchy, false);
+			ConfigurationItem configItem;
+			if (configItems != null) {
+				configItem = configItems.get(0);
+				List<NameValuePair> nvps = configItem.getNameValuePairs();
+				if (nvps != null && nvps.size() > 0) {
+					processImports = true;
+				}
+			}
+			
+			setTheJenaModel(configMgr.getOntModel(publicUri, processImports ? Scope.INCLUDEIMPORTS : Scope.LOCALONLY));
 			setTheJenaModelWithImports(configMgr.getOntModel(publicUri, Scope.INCLUDEIMPORTS));
 		}
 		setAnchor(startNode);

--- a/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl/src/com/ge/research/sadl/builder/ConfigurationManagerForIDE.java
+++ b/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl/src/com/ge/research/sadl/builder/ConfigurationManagerForIDE.java
@@ -68,6 +68,7 @@ import com.ge.research.sadl.model.ConceptName.ConceptType;
 import com.ge.research.sadl.model.SadlSerializationFormat;
 import com.ge.research.sadl.processing.SadlConstants;
 import com.ge.research.sadl.reasoner.ConfigurationException;
+import com.ge.research.sadl.reasoner.ConfigurationItem;
 import com.ge.research.sadl.reasoner.ConfigurationManagerForEditing;
 import com.ge.research.sadl.reasoner.IReasoner;
 import com.ge.research.sadl.reasoner.ITranslator;
@@ -87,6 +88,9 @@ import com.google.inject.Inject;
  */
 public class ConfigurationManagerForIDE extends ConfigurationManagerForEditing implements IConfigurationManagerForIDE {
 	
+	private static final String[] PROJECT_DEPENDENCIES_CATEGORY = {"ProjectDependencies"};
+	private static final String pdependsOn = "pDependsOn";
+
 	public static class Provider implements javax.inject.Provider<IConfigurationManagerForIDE> {
 		private String modelFolder;
 		private String repoType = ConfigurationManagerForIDE.getOWLFormat();
@@ -1300,4 +1304,72 @@ public class ConfigurationManagerForIDE extends ConfigurationManagerForEditing i
 		}
 	}
 
+	@Override
+	public boolean addProjectDependencies(List<java.net.URI> projectDependencies) throws ConfigurationException {
+		File mfp = getModelFolderPath();
+		String thisProjPath = mfp.getParent();
+		for (java.net.URI pduri : projectDependencies) {
+			System.err.println("Project '" + thisProjPath + "' depends on '" + pduri.toString());
+		}
+		String[] categoryHierarchy = PROJECT_DEPENDENCIES_CATEGORY;
+		
+		List<ConfigurationItem> configItems = getConfiguration(categoryHierarchy, false);
+		ConfigurationItem configItem;
+		if (configItems != null) {
+			configItem = configItems.get(0);
+		}
+		else {
+			configItem = new ConfigurationItem(categoryHierarchy);
+		}
+		configItem.clearNameValuePairs();
+		if (projectDependencies != null) {
+			for (java.net.URI pduri : projectDependencies) {
+				ConfigurationItem.NameValuePair nv = configItem.new NameValuePair(pdependsOn, pduri.getPath());
+				configItem.addNameValuePair(nv);
+			}
+		}
+		updateConfiguration(configItem);;
+		saveConfiguration();
+		return true;
+	}
+	
+	private List<String> getProjectDependencies() throws ConfigurationException {
+		List<ConfigurationItem> config = getConfiguration(PROJECT_DEPENDENCIES_CATEGORY, false);
+		List<Object> previousQueries = null;
+		if (config != null && config.size() > 0) {
+			previousQueries  = config.get(0).getAllValuesOfName(pdependsOn);
+		}
+		if (previousQueries != null && previousQueries.size() >= 1) {
+			List<String> dependsOnProjects = new ArrayList<String>();
+			for (Object dependsOn : previousQueries) {
+				dependsOnProjects.add(dependsOn.toString());
+			}
+		}
+		return null;
+	}
+
+	@Override
+	public String getAltUrlFromPublicUri(String publicUri) throws ConfigurationException {
+		try {
+			return super.getAltUrlFromPublicUri(publicUri);
+		}
+		catch (ConfigurationException e) {
+			List<String> pds = getProjectDependencies();
+			if (pds != null) {
+				for (String pd : pds) {
+					String dpModelFolderPath = pd + "/OwlModels";
+					ConfigurationManagerForIDE pdConfigMgr = ConfigurationManagerForIdeFactory.getConfigurationManagerForIDE(dpModelFolderPath, null);
+					if (pdConfigMgr != null) {
+						try {
+							return pdConfigMgr.getAltUrlFromPublicUri(publicUri);
+						}
+						catch (ConfigurationException e2) {
+							// do nothing so loop continues
+						}
+					}
+				}
+			}
+			throw e;
+		}
+	}
 }

--- a/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl/src/com/ge/research/sadl/builder/ConfigurationManagerForIDE.java
+++ b/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl/src/com/ge/research/sadl/builder/ConfigurationManagerForIDE.java
@@ -69,6 +69,7 @@ import com.ge.research.sadl.model.SadlSerializationFormat;
 import com.ge.research.sadl.processing.SadlConstants;
 import com.ge.research.sadl.reasoner.ConfigurationException;
 import com.ge.research.sadl.reasoner.ConfigurationItem;
+import com.ge.research.sadl.reasoner.ConfigurationItem.NameValuePair;
 import com.ge.research.sadl.reasoner.ConfigurationManagerForEditing;
 import com.ge.research.sadl.reasoner.IReasoner;
 import com.ge.research.sadl.reasoner.ITranslator;
@@ -88,9 +89,6 @@ import com.google.inject.Inject;
  */
 public class ConfigurationManagerForIDE extends ConfigurationManagerForEditing implements IConfigurationManagerForIDE {
 	
-	private static final String[] PROJECT_DEPENDENCIES_CATEGORY = {"ProjectDependencies"};
-	private static final String pdependsOn = "pDependsOn";
-
 	public static class Provider implements javax.inject.Provider<IConfigurationManagerForIDE> {
 		private String modelFolder;
 		private String repoType = ConfigurationManagerForIDE.getOWLFormat();
@@ -1321,15 +1319,35 @@ public class ConfigurationManagerForIDE extends ConfigurationManagerForEditing i
 		else {
 			configItem = new ConfigurationItem(categoryHierarchy);
 		}
-		configItem.clearNameValuePairs();
-		if (projectDependencies != null) {
-			for (java.net.URI pduri : projectDependencies) {
-				ConfigurationItem.NameValuePair nv = configItem.new NameValuePair(pdependsOn, pduri.getPath());
-				configItem.addNameValuePair(nv);
+		boolean bChanged = false;
+		List<NameValuePair> nvps = configItem.getNameValuePairs();
+		if (nvps.size() != projectDependencies.size()) {
+			bChanged = true;
+		}
+		if (!bChanged) {
+			List<String> dpStrings = new ArrayList<String>();
+			for (java.net.URI pd : projectDependencies) {
+				dpStrings.add(pd.toString());
+			}
+			for (NameValuePair nvp : nvps) {
+				String nvpv = nvp.getValue().toString();
+				if (!dpStrings.contains(nvpv)) {
+					bChanged = true;
+					break;
+				}
 			}
 		}
-		updateConfiguration(configItem);;
-		saveConfiguration();
+		if (bChanged) {
+			configItem.clearNameValuePairs();
+			if (projectDependencies != null) {
+				for (java.net.URI pduri : projectDependencies) {
+					ConfigurationItem.NameValuePair nv = configItem.new NameValuePair(pdependsOn, pduri.toString());
+					configItem.addNameValuePair(nv);
+				}
+			}
+			updateConfiguration(configItem);;
+			saveConfiguration();
+		}
 		return true;
 	}
 	

--- a/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl/src/com/ge/research/sadl/builder/ConfigurationManagerForIDE.java
+++ b/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl/src/com/ge/research/sadl/builder/ConfigurationManagerForIDE.java
@@ -1304,10 +1304,12 @@ public class ConfigurationManagerForIDE extends ConfigurationManagerForEditing i
 
 	@Override
 	public boolean addProjectDependencies(List<java.net.URI> projectDependencies) throws ConfigurationException {
-		File mfp = getModelFolderPath();
-		String thisProjPath = mfp.getParent();
-		for (java.net.URI pduri : projectDependencies) {
-			System.err.println("Project '" + thisProjPath + "' depends on '" + pduri.toString());
+		if (logger.isDebugEnabled()) {
+			File mfp = getModelFolderPath();
+			String thisProjPath = mfp.getParent();
+			for (java.net.URI pduri : projectDependencies) {
+				logger.debug("Project '" + thisProjPath + "' depends on '" + pduri.toString());
+			}
 		}
 		String[] categoryHierarchy = PROJECT_DEPENDENCIES_CATEGORY;
 		
@@ -1351,43 +1353,4 @@ public class ConfigurationManagerForIDE extends ConfigurationManagerForEditing i
 		return true;
 	}
 	
-	private List<String> getProjectDependencies() throws ConfigurationException {
-		List<ConfigurationItem> config = getConfiguration(PROJECT_DEPENDENCIES_CATEGORY, false);
-		List<Object> previousQueries = null;
-		if (config != null && config.size() > 0) {
-			previousQueries  = config.get(0).getAllValuesOfName(pdependsOn);
-		}
-		if (previousQueries != null && previousQueries.size() >= 1) {
-			List<String> dependsOnProjects = new ArrayList<String>();
-			for (Object dependsOn : previousQueries) {
-				dependsOnProjects.add(dependsOn.toString());
-			}
-		}
-		return null;
-	}
-
-	@Override
-	public String getAltUrlFromPublicUri(String publicUri) throws ConfigurationException {
-		try {
-			return super.getAltUrlFromPublicUri(publicUri);
-		}
-		catch (ConfigurationException e) {
-			List<String> pds = getProjectDependencies();
-			if (pds != null) {
-				for (String pd : pds) {
-					String dpModelFolderPath = pd + "/OwlModels";
-					ConfigurationManagerForIDE pdConfigMgr = ConfigurationManagerForIdeFactory.getConfigurationManagerForIDE(dpModelFolderPath, null);
-					if (pdConfigMgr != null) {
-						try {
-							return pdConfigMgr.getAltUrlFromPublicUri(publicUri);
-						}
-						catch (ConfigurationException e2) {
-							// do nothing so loop continues
-						}
-					}
-				}
-			}
-			throw e;
-		}
-	}
 }

--- a/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl/src/com/ge/research/sadl/builder/IConfigurationManagerForIDE.java
+++ b/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl/src/com/ge/research/sadl/builder/IConfigurationManagerForIDE.java
@@ -119,6 +119,14 @@ public interface IConfigurationManagerForIDE extends IConfigurationManagerForEdi
 	 */
 	public abstract boolean addJenaMapping(String publicUri, String altUrl)
 			throws IOException, URISyntaxException, ConfigurationException;
+	
+	/**
+	 * Method to record dependencies of this Project (KBase) on other Projects (KBases)
+	 * @param dependencies
+	 * @return
+	 * @throws ConfigurationException 
+	 */
+	public abstract boolean addProjectDependencies(List<java.net.URI> dependencies) throws ConfigurationException;
 
 	public abstract boolean deleteModel(String publicUri)
 			throws ConfigurationException, IOException, URISyntaxException;

--- a/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl/src/com/ge/research/sadl/utils/EclipseSadlProjectHelper.xtend
+++ b/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl/src/com/ge/research/sadl/utils/EclipseSadlProjectHelper.xtend
@@ -17,7 +17,6 @@
  ***********************************************************************/
 package com.ge.research.sadl.utils
 
-import com.ge.research.sadl.utils.SadlProjectHelper
 import com.google.common.base.Preconditions
 import java.net.URI
 import java.nio.file.Path
@@ -78,6 +77,18 @@ class EclipseSadlProjectHelper implements SadlProjectHelper {
 			throw new IllegalArgumentException('''The resource does not exits for the «uri» URI. Converted path was: «path».''');
 		}
 		return Paths.get(resource.locationURI);
+	}
+
+	override getReferencedProjectURIs(URI uri) {
+		val emfRootUri = uri.toEmfUri
+		if (emfRootUri.segmentCount < 2) {
+			return emptyList
+		}
+		val project = getProject(emfRootUri.trimSegments(emfRootUri.segmentCount() - 2))
+		if (project === null || !project.accessible) {
+			return emptyList
+		}
+		return project.description.referencedProjects.filterNull.filter[accessible].map[locationURI].filterNull.toList
 	}
 
 	protected def checkUri(URI uri) {

--- a/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl/src/com/ge/research/sadl/utils/SadlIdeProjectHelper.xtend
+++ b/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl/src/com/ge/research/sadl/utils/SadlIdeProjectHelper.xtend
@@ -17,12 +17,16 @@
  ***********************************************************************/
 package com.ge.research.sadl.utils
 
-import com.ge.research.sadl.utils.ResourceManager
-import com.ge.research.sadl.utils.SadlProjectHelper
 import com.google.common.base.Preconditions
+import java.io.ByteArrayInputStream
+import java.io.InputStream
 import java.net.URI
+import java.nio.charset.StandardCharsets
 import java.nio.file.Files
+import java.nio.file.Path
 import java.nio.file.Paths
+import java.util.List
+import javax.xml.parsers.DocumentBuilderFactory
 
 /**
  * Helper class for validating and locating SADL projects on the file system.
@@ -86,4 +90,56 @@ class SadlIdeProjectHelper implements SadlProjectHelper {
 		return uri;
 	}
 
+	override getReferencedProjectURIs(URI uri) {
+		val rootUri = uri.getRoot
+		val rootPath = rootUri.toPath
+		val dotProjectPath = rootPath.resolve(DOT_PROJECT)
+		val projectNames = new ReferencedProjectsParser().parse(dotProjectPath)
+		projectNames.map[rootPath.resolve(it)].map[toFile].filter[exists && directory && canRead].map[toURI].toList
+	}
+
+}
+
+class ReferencedProjectsParser {
+
+	def List<String> parse(Path path) {
+		var InputStream is = null
+		try {
+			is = path.toUri.toURL.openStream
+			return parse(is)
+		} finally {
+			if (is !== null) {
+				is.close
+			}
+		}
+	}
+
+	def List<String> parse(CharSequence content) {
+		return parse(new ByteArrayInputStream(content.toString().getBytes(StandardCharsets.UTF_8)))
+	}
+
+	def List<String> parse(InputStream is) {
+		try {
+			val projectNames = newLinkedHashSet;
+			val factory = DocumentBuilderFactory.newInstance()
+			val builder = factory.newDocumentBuilder()
+			val doc = builder.parse(is)
+			doc.getDocumentElement().normalize()
+			val projectNode = doc.getElementsByTagName('project')
+			for (var i = 0; i < projectNode.length; i++) {
+				val project = projectNode.item(i)
+				val children = project.childNodes
+				for (var j = 0; j < children.length; j++) {
+					val projectName = children.item(j).nodeValue
+					if (!projectName.nullOrEmpty) {
+						projectNames.add(projectName)
+					}
+				}
+			}
+			return projectNames.toList
+		} catch (Exception e) {
+			e.printStackTrace
+			return newArrayList
+		}
+	}
 }

--- a/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl/src/com/ge/research/sadl/utils/SadlProjectHelper.xtend
+++ b/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl/src/com/ge/research/sadl/utils/SadlProjectHelper.xtend
@@ -22,7 +22,9 @@ import com.google.inject.Inject
 import com.google.inject.Singleton
 import java.net.URI
 import java.nio.file.Path
+import java.util.List
 import org.eclipse.emf.common.EMFPlugin
+import org.eclipse.emf.ecore.resource.Resource
 import org.eclipse.xtext.util.internal.Log
 
 import static extension com.google.common.base.Strings.nullToEmpty
@@ -58,6 +60,16 @@ interface SadlProjectHelper {
 	 */
 	def Path toPath(URI uri);
 
+	/**
+	 * Returns the URI of the projects referenced by the project give project URI.
+	 * The URIs point to accessible projects. If the project does not reference any projects,
+	 * returns an empty list.
+	 */
+	def List<URI> getReferencedProjectURIs(URI uri);
+	def List<URI> getReferencedProjectURIs(Resource resource) {
+		return getReferencedProjectURIs(new URI(resource.URI.toString()));
+	}
+
 	@Log
 	@Singleton
 	class DispatchingProjectHelper implements SadlProjectHelper {
@@ -81,6 +93,10 @@ interface SadlProjectHelper {
 
 		override toPath(URI uri) {
 			return uri.dispatch.toPath(uri);
+		}
+		
+		override getReferencedProjectURIs(URI uri) {
+			return uri.dispatch.getReferencedProjectURIs(uri);
 		}
 
 		override toUri(Path path) {
@@ -121,6 +137,13 @@ interface SadlProjectHelper {
 
 		override toPath(URI uri) {
 			return null;
+		}
+
+		/**
+		 * Always empty. Project dependencies are not supported in headless mode.
+		 */
+		override getReferencedProjectURIs(URI uri) {
+			return newArrayList;
 		}
 
 	}

